backend/executor.go

```diff
@@ -360,25 +373,29 @@ func (g *grpcExecutor) GetWorkItems(req *protos.GetWorkItemsRequest, stream prot
 	}
 }
 
-func (g *grpcExecutor) sendWorkItem(stream protos.TaskHubSidecarService_GetWorkItemsServer, wi *protos.WorkItem) error {
+func (g *grpcExecutor) sendWorkItem(stream protos.TaskHubSidecarService_GetWorkItemsServer, wi *protos.WorkItem,
+	ch chan *protos.WorkItem, errCh chan error,
+) error {
+	select {
+	case <-stream.Context().Done():
+		return stream.Context().Err()
+	case ch <- wi:
```

Comment:
I think this all should be after the line 390, you should use the context that includes the optional timeout for the send as well

Reply:
I don't think so, as this check is only important for when the stream has been closed. If it is not closed, the ch should be read from immediately, because the loop this is in is a singleton, and in-turn, always waiting for the errCh result. If an error happens at all, everything errors out.

Reply:
This is the root of the problem bc `stream.Send` does not take a ctx or deadline, so the deadline handling needs to be outside of the same go routine.